### ISSUE TITLE
Workflow not marking itself as completed when finished

### ISSUE THREAD

**Issue by tim-sendible** [argoproj/argo-workflows#5481]
## Summary 

My workflow shows as running even though all nodes have completed. This happens regardless of whether or not the nodes pass or some fail.

Other information that may be useful:

- We believe the issue is something to do with the retryStrategy. I can **only** replicate this issue when I set a retry strategy.
- (most of) My workflows have an exit handler. These are not triggered when this issue occurs, but are triggered when it doesn't.
- This issue occurs each time and for multiple different workflow configurations.
- When this bug is triggered, the workflow phase is stuck on 'running' according to the labels.

## Diagnostics

What Kubernetes provider are you using? EKS 1.19

What version of Argo Workflows are you running? 3.0 rc8

```yaml
Paste a workflow that reproduces the bug, including status:

Each workflow we have produces this error. Here's (what I believe to be) the relevant pieces for you:
    - template:
        name: github-workflow-trigger
        k8s:
          group: argoproj.io
          version: v1alpha1
          resource: workflows
          operation: create
          source:
            resource:
              apiVersion: argoproj.io/v1alpha1
              kind: Workflow
              metadata:
                generateName: classic-pr-
                namespace: ci
              spec:
                # Will timeout in 1 hour
                activeDeadlineSeconds: 3600
                # Will delete the workflow 1 hour after completion
                ttlStrategy:
                  secondsAfterCompletion: 3600
                # Delete volumes on workflow completion
                volumeClaimGC:
                  strategy: OnWorkflowCompletion
                # Delete a pod immediately when the pod is successful
                podGC:
                  strategy: OnWorkflowCompletion
                # retryStrategy:
                #   retryPolicy: OnError
                #   limit: 3
                onExit: exit-handler
                entrypoint: main
```

```
Paste the logs from the workflow controller (note these logs go in asc order):
	
2021-03-22 16:05:12	
time="2021-03-22T16:05:12.867Z" level=info msg="Patch events 200"
	
2021-03-22 16:05:12	
time="2021-03-22T16:05:12.867Z" level=info msg="Workflow update successful" namespace=ci phase=Running resourceVersion=31835797 workflow=classic-pr-cbvmf
	
2021-03-22 16:05:12	
time="2021-03-22T16:05:12.856Z" level=info msg="Update workflows 200"
	
2021-03-22 16:05:12	
time="2021-03-22T16:05:12.855Z" level=info msg="Patch events 200"
	
2021-03-22 16:05:12	
time="2021-03-22T16:05:12.843Z" level=info msg="Patch events 200"
	
2021-03-22 16:05:12	
time="2021-03-22T16:05:12.832Z" level=info msg="Patch events 200"
	
2021-03-22 16:05:12	
time="2021-03-22T16:05:12.825Z" level=debug msg="Log changes patch: {\"status\":{\"nodes\":{\"classic-pr-cbvmf\":{\"finishedAt\":\"2021-03-22T16:05:12Z\",\"phase\":\"Succeeded\"},\"classic-pr-cbvmf-2055809241\":{\"finishedAt\":\"2021-03-22T16:05:04Z\",\"phase\":\"Succeeded\",\"resourcesDuration\":{\"cpu\":0,\"memory\":0}},\"classic-pr-cbvmf-2952901088\":{\"finishedAt\":\"2021-03-22T16:05:12Z\",\"phase\":\"Succeeded\"},\"classic-pr-cbvmf-3716595798\":{\"finishedAt\":\"2021-03-22T16:05:12Z\",\"outboundNodes\":[\"classic-pr-cbvmf-581260995\",\"classic-pr-cbvmf-4082496012\",\"classic-pr-cbvmf-1984848737\"],\"phase\":\"Succeeded\"},\"classic-pr-cbvmf-4166644962\":{\"finishedAt\":\"2021-03-22T16:05:12Z\",\"phase\":\"Succeeded\"},\"classic-pr-cbvmf-581260995\":{\"finishedAt\":\"2021-03-22T16:05:12Z\",\"outboundNodes\":[\"classic-pr-cbvmf-2055809241\",\"classic-pr-cbvmf-683031197\",\"classic-pr-cbvmf-1314800059\",\"classic-pr-cbvmf-2674077415\"],\"phase\":\"Succeeded\"}}}}"
	
2021-03-22 16:05:12	
time="2021-03-22T16:05:12.822Z" level=info msg="Patch events 200"
	
2021-03-22 16:05:12	
time="2021-03-22T16:05:12.817Z" level=debug msg="Event(v1.ObjectReference{Kind:\"Workflow\", Namespace:\"ci\", Name:\"classic-pr-cbvmf\", UID:\"eaee0f10-8248-4137-a97f-68366ea68623\", APIVersion:\"argoproj.io/v1alpha1\", ResourceVersion:\"31835602\", FieldPath:\"\"}): type: 'Normal' reason: 'WorkflowNodeSucceeded' Succeeded node classic-pr-cbvmf"
	
2021-03-22 16:05:12	
time="2021-03-22T16:05:12.817Z" level=info msg="node classic-pr-cbvmf finished: 2021-03-22 16:05:12.817169236 +0000 UTC" namespace=ci workflow=classic-pr-cbvmf
	
2021-03-22 16:05:12	
time="2021-03-22T16:05:12.817Z" level=info msg="node classic-pr-cbvmf phase Running -> Succeeded" namespace=ci workflow=classic-pr-cbvmf
	
2021-03-22 16:05:12	
time="2021-03-22T16:05:12.817Z" level=debug msg="Executing node classic-pr-cbvmf of Retry is Running" namespace=ci workflow=classic-pr-cbvmf
	
2021-03-22 16:05:12	
time="2021-03-22T16:05:12.816Z" level=debug msg="Event(v1.ObjectReference{Kind:\"Workflow\", Namespace:\"ci\", Name:\"classic-pr-cbvmf\", UID:\"eaee0f10-8248-4137-a97f-68366ea68623\", APIVersion:\"argoproj.io/v1alpha1\", ResourceVersion:\"31835602\", FieldPath:\"\"}): type: 'Normal' reason: 'WorkflowNodeSucceeded' Succeeded node classic-pr-cbvmf(0)"
	
2021-03-22 16:05:12	
time="2021-03-22T16:05:12.816Z" level=debug msg="Getting the template by name" base="*v1alpha1.Workflow (namespace=ci,name=classic-pr-cbvmf)" depth=0 tmpl="*v1alpha1.WorkflowStep (main)"
	
2021-03-22 16:05:12	
time="2021-03-22T16:05:12.816Z" level=debug msg="Getting the template" base="*v1alpha1.Workflow (namespace=ci,name=classic-pr-cbvmf)" depth=0 tmpl="*v1alpha1.WorkflowStep (main)"
	
2021-03-22 16:05:12	
time="2021-03-22T16:05:12.816Z" level=debug msg="Resolving the template" base="*v1alpha1.Workflow (namespace=ci,name=classic-pr-cbvmf)" depth=0 tmpl="*v1alpha1.WorkflowStep (main)"
	
2021-03-22 16:05:12	
time="2021-03-22T16:05:12.816Z" level=debug msg="Evaluating node classic-pr-cbvmf: template: *v1alpha1.WorkflowStep (main), boundaryID: " namespace=ci workflow=classic-pr-cbvmf
	
2021-03-22 16:05:12	
time="2021-03-22T16:05:12.816Z" level=info msg="Checking daemoned children of classic-pr-cbvmf-3716595798" namespace=ci workflow=classic-pr-cbvmf
	
2021-03-22 16:05:12	
time="2021-03-22T16:05:12.816Z" level=info msg="node classic-pr-cbvmf-3716595798 finished: 2021-03-22 16:05:12.816455174 +0000 UTC" namespace=ci workflow=classic-pr-cbvmf
	
2021-03-22 16:05:12	
time="2021-03-22T16:05:12.816Z" level=info msg="node classic-pr-cbvmf-3716595798 phase Running -> Succeeded" namespace=ci workflow=classic-pr-cbvmf
	
2021-03-22 16:05:12	
time="2021-03-22T16:05:12.816Z" level=info msg="Outbound nodes of classic-pr-cbvmf-3716595798 set to [classic-pr-cbvmf-581260995 classic-pr-cbvmf-4082496012 classic-pr-cbvmf-1984848737]" namespace=ci workflow=classic-pr-cbvmf
	
2021-03-22 16:05:12	
time="2021-03-22T16:05:12.816Z" level=debug msg="Found stored template" base="*v1alpha1.Workflow (namespace=ci,name=classic-pr-cbvmf)" depth=0 tmpl="*v1alpha1.DAGTask (sonarscanner-classic/main)"
	
2021-03-22 16:05:12	
time="2021-03-22T16:05:12.816Z" level=debug msg="Resolving the template" base="*v1alpha1.Workflow (namespace=ci,name=classic-pr-cbvmf)" depth=0 tmpl="*v1alpha1.DAGTask (sonarscanner-classic/main)"
	
2021-03-22 16:05:12	
time="2021-03-22T16:05:12.815Z" level=debug msg="Found stored template" base="*v1alpha1.Workflow (namespace=ci,name=classic-pr-cbvmf)" depth=0 tmpl="*v1alpha1.DAGTask (pr-merge-checker/main)"
	
2021-03-22 16:05:12	
time="2021-03-22T16:05:12.815Z" level=debug msg="Event(v1.ObjectReference{Kind:\"Workflow\", Namespace:\"ci\", Name:\"classic-pr-cbvmf\", UID:\"eaee0f10-8248-4137-a97f-68366ea68623\", APIVersion:\"argoproj.io/v1alpha1\", ResourceVersion:\"31835602\", FieldPath:\"\"}): type: 'Normal' reason: 'WorkflowNodeSucceeded' Succeeded node classic-pr-cbvmf(0).classic-unittest-classic"
	
2021-03-22 16:05:12	
time="2021-03-22T16:05:12.815Z" level=debug msg="Resolving the template" base="*v1alpha1.Workflow (namespace=ci,name=classic-pr-cbvmf)" depth=0 tmpl="*v1alpha1.DAGTask (pr-merge-checker/main)"
	
2021-03-22 16:05:12	
time="2021-03-22T16:05:12.815Z" level=info msg="node classic-pr-cbvmf-2952901088 finished: 2021-03-22 16:05:12.815325644 +0000 UTC" namespace=ci workflow=classic-pr-cbvmf
	
2021-03-22 16:05:12	
time="2021-03-22T16:05:12.815Z" level=info msg="node classic-pr-cbvmf-2952901088 phase Running -> Succeeded" namespace=ci workflow=classic-pr-cbvmf
	
2021-03-22 16:05:12	
time="2021-03-22T16:05:12.815Z" level=debug msg="Executing node classic-pr-cbvmf(0).classic-unittest-classic of Retry is Running" namespace=ci workflow=classic-pr-cbvmf
	
2021-03-22 16:05:12	
time="2021-03-22T16:05:12.814Z" level=info msg="Patch events 200"

```

  |  
-- | --

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by simster7** [comment]
I'll take a look

**Comment by simster7** [comment]
> This issue occurs each time and for multiple different workflow configurations.

@tim-sendible Can you please attach a minimal workflow that reproduces this problem? The workflow you attach is not complete and it's not clear how the issue manifests from it

**Comment by simster7** [comment]
Also, perhaps the thing that would most help me debug this issue is if you would include a version of the "live" workflow that suffers from this issue. You can get the "live" workflow by running `kubectl get workflow -o yaml <WF_NAME>`.

**Comment by tim-sendible** [comment]
@simster7 It occurs on literally any workflow we run as long as the retry strategy is enabled. 
However if you like I will try and provide a sanitised workflow tomorrow.

**Comment by simster7** [comment]
Yes, please that would be useful as I haven't been able to reproduce this.

Some other questions to help debugging: does this occur only on v3.0.0-rc8 or does it also occur in other RCs? Does it occur in v2.12.*?

**Comment by tim-sendible** [comment]
I have only been using workflows for 2 weeks. It was really only since using rc8 that I have been in a position to notice the hanging. 

So there is a slim chance it's only occurred in rc8. I have only used rc6 and rc7 and have not used v2 for anything other than proving the installation process.

**Comment by simster7** [comment]
Great, thanks for the info. I'll keep trying to replicate, otherwise the sanitized "live" workflows would be most useful

**Comment by tim-sendible** [comment]
Here's a sanitised example:

```
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  creationTimestamp: "2021-03-23T06:22:21Z"
  generateName: never-completes-
  generation: 5
  labels:
    workflows.argoproj.io/creator: Cgg2OTQ2MTQ0ORIGZ2l0aHVi
    workflows.argoproj.io/phase: Running
  managedFields:
  - apiVersion: argoproj.io/v1alpha1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:generateName: {}
        f:labels:
          .: {}
          f:workflows.argoproj.io/creator: {}
      f:spec:
        .: {}
        f:activeDeadlineSeconds: {}
        f:arguments: {}
        f:entrypoint: {}
        f:nodeSelector: {}
        f:podGC: {}
        f:retryStrategy: {}
        f:serviceAccountName: {}
        f:templates: {}
        f:ttlStrategy: {}
        f:volumeClaimGC: {}
        f:volumeClaimTemplates: {}
        f:volumes: {}
      f:status:
        .: {}
        f:finishedAt: {}
    manager: argo
    operation: Update
    time: "2021-03-23T06:22:21Z"
  - apiVersion: argoproj.io/v1alpha1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:labels:
          f:workflows.argoproj.io/phase: {}
      f:status:
        f:artifactRepositoryRef: {}
        f:conditions: {}
        f:nodes: {}
        f:persistentVolumeClaims: {}
        f:phase: {}
        f:progress: {}
        f:resourcesDuration: {}
        f:startedAt: {}
    manager: workflow-controller
    operation: Update
    time: "2021-03-23T06:22:31Z"
  name: never-completes-9ggbz
  namespace: ci
  resourceVersion: "32181340"
  selfLink: /apis/argoproj.io/v1alpha1/namespaces/ci/workflows/never-completes-9ggbz
  uid: 6d2d6363-3cfa-498d-b31a-f84bdd110bfd
spec:
  activeDeadlineSeconds: 3600
  arguments: {}
  entrypoint: main
  nodeSelector:
    eks.amazonaws.com/nodegroup: spot-nodes
  podGC:
    strategy: OnWorkflowCompletion
  retryStrategy:
    limit: 3
    retryPolicy: OnError
  serviceAccountName: argo-workflows
  templates:
  - dag:
      tasks:
      - arguments: {}
        name: random-worker
        template: random-worker
        withItems:
        - 1
        - 2
        - 3
        - 4
        - 5
    inputs: {}
    metadata: {}
    name: main
    outputs: {}
  - container:
      args:
      - echo "Wibble" && sleep 5 && echo "Dun"
      command:
      - sh
      - -c
      image: docker.sendilab.net/proxy/library/alpine
      imagePullPolicy: Always
      name: ""
      resources:
        requests:
          cpu: "1"
          memory: 700Mi
    inputs: {}
    metadata: {}
    name: random-worker
    outputs: {}
  ttlStrategy:
    secondsAfterCompletion: 3600
  volumeClaimGC:
    strategy: OnWorkflowCompletion
  volumeClaimTemplates:
  - metadata:
      creationTimestamp: null
      name: workdir
    spec:
      accessModes:
      - ReadWriteMany
      resources:
        requests:
          storage: 1Gi
      storageClassName: nfs
    status: {}
  volumes:
  - emptyDir: {}
    name: worker1
  - emptyDir: {}
    name: worker2
  - emptyDir: {}
    name: worker3
  - emptyDir: {}
    name: worker4
  - emptyDir: {}
    name: worker5
status:
  artifactRepositoryRef:
    default: true
  conditions:
  - status: "False"
    type: PodRunning
  finishedAt: null
  nodes:
    never-completes-9ggbz:
      children:
      - never-completes-9ggbz-82303199
      displayName: never-completes-9ggbz
      finishedAt: "2021-03-23T06:22:52Z"
      id: never-completes-9ggbz
      name: never-completes-9ggbz
      phase: Succeeded
      progress: 5/5
      resourcesDuration:
        cpu: 55
        memory: 205
      startedAt: "2021-03-23T06:22:21Z"
      templateName: main
      templateScope: local/never-completes-9ggbz
      type: Retry
    never-completes-9ggbz-82303199:
      children:
      - never-completes-9ggbz-909830033
      displayName: never-completes-9ggbz(0)
      finishedAt: "2021-03-23T06:22:52Z"
      id: never-completes-9ggbz-82303199
      name: never-completes-9ggbz(0)
      outboundNodes:
      - never-completes-9ggbz-1514290502
      - never-completes-9ggbz-3909440240
      - never-completes-9ggbz-663838326
      - never-completes-9ggbz-385440284
      - never-completes-9ggbz-2512596758
      phase: Succeeded
      progress: 5/5
      resourcesDuration:
        cpu: 55
        memory: 205
      startedAt: "2021-03-23T06:22:21Z"
      templateName: main
      templateScope: local/never-completes-9ggbz
      type: DAG
    never-completes-9ggbz-385440284:
      boundaryID: never-completes-9ggbz-82303199
      displayName: random-worker(3:4)(0)
      finishedAt: "2021-03-23T06:22:36Z"
      hostNodeName: ip-172-21-2-59.ec2.internal
      id: never-completes-9ggbz-385440284
      name: never-completes-9ggbz(0).random-worker(3:4)(0)
      phase: Succeeded
      progress: 1/1
      resourcesDuration:
        cpu: 11
        memory: 41
      startedAt: "2021-03-23T06:22:21Z"
      templateName: random-worker
      templateScope: local/never-completes-9ggbz
      type: Pod
    never-completes-9ggbz-663838326:
      boundaryID: never-completes-9ggbz-82303199
      displayName: random-worker(2:3)(0)
      finishedAt: "2021-03-23T06:22:29Z"
      id: never-completes-9ggbz-663838326
      name: never-completes-9ggbz(0).random-worker(2:3)(0)
      phase: Succeeded
      progress: 1/1
      resourcesDuration:
        cpu: 11
        memory: 41
      startedAt: "2021-03-23T06:22:21Z"
      templateName: random-worker
      templateScope: local/never-completes-9ggbz
      type: Pod
    never-completes-9ggbz-808390395:
      boundaryID: never-completes-9ggbz-82303199
      children:
      - never-completes-9ggbz-663838326
      displayName: random-worker(2:3)
      finishedAt: "2021-03-23T06:22:31Z"
      id: never-completes-9ggbz-808390395
      name: never-completes-9ggbz(0).random-worker(2:3)
      phase: Succeeded
      progress: 1/1
      resourcesDuration:
        cpu: 11
        memory: 41
      startedAt: "2021-03-23T06:22:21Z"
      templateName: random-worker
      templateScope: local/never-completes-9ggbz
      type: Retry
    never-completes-9ggbz-909830033:
      boundaryID: never-completes-9ggbz-82303199
      children:
      - never-completes-9ggbz-2725971403
      - never-completes-9ggbz-1368469385
      - never-completes-9ggbz-808390395
      - never-completes-9ggbz-1383147341
      - never-completes-9ggbz-3708144667
      displayName: random-worker
      finishedAt: "2021-03-23T06:22:52Z"
      id: never-completes-9ggbz-909830033
      name: never-completes-9ggbz(0).random-worker
      phase: Succeeded
      progress: 5/5
      resourcesDuration:
        cpu: 55
        memory: 205
      startedAt: "2021-03-23T06:22:21Z"
      templateName: random-worker
      templateScope: local/never-completes-9ggbz
      type: TaskGroup
    never-completes-9ggbz-1368469385:
      boundaryID: never-completes-9ggbz-82303199
      children:
      - never-completes-9ggbz-3909440240
      displayName: random-worker(1:2)
      finishedAt: "2021-03-23T06:22:31Z"
      id: never-completes-9ggbz-1368469385
      name: never-completes-9ggbz(0).random-worker(1:2)
      phase: Succeeded
      progress: 1/1
      resourcesDuration:
        cpu: 11
        memory: 41
      startedAt: "2021-03-23T06:22:21Z"
      templateName: random-worker
      templateScope: local/never-completes-9ggbz
      type: Retry
    never-completes-9ggbz-1383147341:
      boundaryID: never-completes-9ggbz-82303199
      children:
      - never-completes-9ggbz-385440284
      displayName: random-worker(3:4)
      finishedAt: "2021-03-23T06:22:41Z"
      id: never-completes-9ggbz-1383147341
      name: never-completes-9ggbz(0).random-worker(3:4)
      phase: Succeeded
      progress: 1/1
      resourcesDuration:
        cpu: 11
        memory: 41
      startedAt: "2021-03-23T06:22:21Z"
      templateName: random-worker
      templateScope: local/never-completes-9ggbz
      type: Retry
    never-completes-9ggbz-1514290502:
      boundaryID: never-completes-9ggbz-82303199
      displayName: random-worker(0:1)(0)
      finishedAt: "2021-03-23T06:22:29Z"
      id: never-completes-9ggbz-1514290502
      name: never-completes-9ggbz(0).random-worker(0:1)(0)
      phase: Succeeded
      progress: 1/1
      resourcesDuration:
        cpu: 11
        memory: 41
      startedAt: "2021-03-23T06:22:21Z"
      templateName: random-worker
      templateScope: local/never-completes-9ggbz
      type: Pod
    never-completes-9ggbz-2512596758:
      boundaryID: never-completes-9ggbz-82303199
      displayName: random-worker(4:5)(0)
      finishedAt: "2021-03-23T06:22:41Z"
      hostNodeName: ip-172-21-3-216.ec2.internal
      id: never-completes-9ggbz-2512596758
      name: never-completes-9ggbz(0).random-worker(4:5)(0)
      phase: Succeeded
      progress: 1/1
      resourcesDuration:
        cpu: 11
        memory: 41
      startedAt: "2021-03-23T06:22:21Z"
      templateName: random-worker
      templateScope: local/never-completes-9ggbz
      type: Pod
    never-completes-9ggbz-2725971403:
      boundaryID: never-completes-9ggbz-82303199
      children:
      - never-completes-9ggbz-1514290502
      displayName: random-worker(0:1)
      finishedAt: "2021-03-23T06:22:31Z"
      id: never-completes-9ggbz-2725971403
      name: never-completes-9ggbz(0).random-worker(0:1)
      phase: Succeeded
      progress: 1/1
      resourcesDuration:
        cpu: 11
        memory: 41
      startedAt: "2021-03-23T06:22:21Z"
      templateName: random-worker
      templateScope: local/never-completes-9ggbz
      type: Retry
    never-completes-9ggbz-3708144667:
      boundaryID: never-completes-9ggbz-82303199
      children:
      - never-completes-9ggbz-2512596758
      displayName: random-worker(4:5)
      finishedAt: "2021-03-23T06:22:52Z"
      id: never-completes-9ggbz-3708144667
      name: never-completes-9ggbz(0).random-worker(4:5)
      phase: Succeeded
      progress: 1/1
      resourcesDuration:
        cpu: 11
        memory: 41
      startedAt: "2021-03-23T06:22:21Z"
      templateName: random-worker
      templateScope: local/never-completes-9ggbz
      type: Retry
    never-completes-9ggbz-3909440240:
      boundaryID: never-completes-9ggbz-82303199
      displayName: random-worker(1:2)(0)
      finishedAt: "2021-03-23T06:22:29Z"
      id: never-completes-9ggbz-3909440240
      name: never-completes-9ggbz(0).random-worker(1:2)(0)
      phase: Succeeded
      progress: 1/1
      resourcesDuration:
        cpu: 11
        memory: 41
      startedAt: "2021-03-23T06:22:21Z"
      templateName: random-worker
      templateScope: local/never-completes-9ggbz
      type: Pod
  persistentVolumeClaims:
  - name: workdir
    persistentVolumeClaim:
      claimName: never-completes-9ggbz-workdir
  phase: Running
  progress: 5/5
  resourcesDuration:
    cpu: 55
    memory: 205
  startedAt: "2021-03-23T06:22:21Z"
```

**Comment by simster7** [comment]
Thanks, I was able to reproduce. Looking for the fix now

**Comment by simster7** [comment]
This has been fixed in [`v3.0.0-rc9`](https://github.com/argoproj/argo-workflows/releases/tag/v3.0.0-rc9)
